invokeai/app/services/download/download_default.py

```diff
@@ -0,0 +1,367 @@
+
+    def __init__(
+        self,
+        max_parallel_dl: int = 5,
+        event_bus: Optional[EventServiceBase] = None,
+        requests_session: Optional[requests.sessions.Session] = None,
+    ):
+        """
+        Initialize DownloadQueue.
+
+        :param max_parallel_dl: Number of simultaneous downloads allowed [5].
+        :param requests_session: Optional requests.sessions.Session object, for unit tests.
+        """
+        self._jobs = {}
+        self._next_job_id = 0
+        self._queue = PriorityQueue()
+        self._worker_pool = set()
+        self._lock = threading.Lock()
+        self._logger = InvokeAILogger.get_logger("DownloadQueueService")
+        self._event_bus = event_bus
+        self._requests = requests_session or requests.Session()
+
+        self._start_workers(max_parallel_dl)
+
+    def __del__(self) -> None:
+        if any(thread.is_alive for thread in self._worker_pool):
+            self._queue.put(STOP_JOB)
+
+    def download(
+        self,
+        source: Union[str, AnyHttpUrl],
+        dest: Union[str, Path],
+        priority: int = 10,
+        access_token: Optional[str] = None,
+        on_start: Optional[DownloadEventHandler] = None,
+        on_progress: Optional[DownloadEventHandler] = None,
+        on_complete: Optional[DownloadEventHandler] = None,
+        on_error: Optional[DownloadEventHandler] = None,
+    ) -> DownloadJob:
+        """Create a download job and return its ID."""
+        with self._lock:
+            id = self._next_job_id
+            self._next_job_id += 1
+            job = DownloadJob(
+                id=id,
+                source=Url(source)
+                if isinstance(source, str)
+                else source,  # maybe a validator in DownloadJob would be better here
+                dest=Path(dest),
+                priority=priority,
+                access_token=access_token,
+            )
+            job.set_callbacks(
+                on_start=on_start,
+                on_progress=on_progress,
+                on_complete=on_complete,
+                on_error=on_error,
+            )
+            self._jobs[id] = job
+            self._queue.put(job)
+        return job
+
+    def join(self):
+        """Wait for all jobs to complete."""
+        self._queue.join()
+
+    def list_jobs(self) -> List[DownloadJob]:
+        """List all the jobs."""
+        return list(self._jobs.values())
+
+    def prune_jobs(self):
+        """Prune completed and errored queue items from the job list."""
+        with self._lock:
+            to_delete = set()
+            try:
+                for job_id, job in self._jobs.items():
+                    if self._in_terminal_state(job):
+                        to_delete.add(job_id)
+                for job_id in to_delete:
+                    del self._jobs[job_id]
+            except KeyError as excp:
+                raise UnknownJobIDException("Unrecognized job") from excp
```

Comment:
minor: `UnknownJobIDException` is used elsewhere to indicate that an invalid job id was provided as an argument. Given that we should never get in this state, I think it makes sense to just allow the `KeyError` to propagate as-is here.

Reply:
Fixed.